src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp

```diff
@@ -125,59 +133,183 @@ struct DestroySharpAlm {
+ * standard mathematical definitions in a nontrivial way. There are a number of
+ * important features to the data storage of the coefficients.
+ * - they are stored as a set of complex values, but each vector of complex
+ *   values is the transform of only the real or imaginary part of the
+ *   collocation data
+ * - because each vector of complex coefficients is related to the transform of
+ *   a set of doubles, only (about) half of the m's are stored (m >= 0), because
+ *   the remaining m modes are determinable by conjugation from the positive m
+ *   modes, given that they represent the transform of a purely real or purely
+ *   imaginary collocation quantity
+ * - they are stored in an l-varies-fastest, triangular representation. To be
+ *   concrete, for an l_max=2, the order of coefficient storage is (l, m):
+ *   [(0, 0), (1, 0), (2, 0), (1, 1), (2, 1), (2, 2)]
+ * - due to the restriction of representing only the transform of real
+ *   quantities, the m=0 modes always have vanishing imaginary component.
+ */
+class CoefficientsMetadata {
  public:
-  explicit Coefficients(size_t l_max) noexcept;
-
-  ~Coefficients() = default;
-  Coefficients() = default;
-  Coefficients(const Coefficients&) = delete;
-  Coefficients(Coefficients&&) = default;
-  Coefficients& operator=(const Coefficients&) = delete;
-  Coefficients& operator=(Coefficients&&) = default;
+  /// An iterator for easily traversing a libsharp-compatible spin-weighted
+  /// spherical harmonic modal representation.
+  /// The `operator*()` returns a `LibsharpCoefficientInfo`, which  contains two
+  /// offsets, `transform_of_real_part_offset` and
+  /// `transform_of_imag_part_offset`, and the `l_max`, `l` and `m` associated
+  /// with the values at those offsets.
+  ///
+  /// \note this currently assumes, as do many of the utilities in this file,
+  /// that the libsharp representation is chosen to be the triangular
+  /// coefficient representation. If alternative representations are desired,
+  /// alterations will be needed.
+  class CoefficientsConstIterator {
+   public:
+    explicit CoefficientsConstIterator(const size_t l_max,
+                                       const size_t start_l = 0,
+                                       const size_t start_m = 0) noexcept
+        : m_{start_m}, l_{start_l}, l_max_{l_max} {}
+
+    LibsharpCoefficientInfo operator*() const noexcept {
+      // permit dereferencing the iterator only if the return represents a
+      // viable location in the coefficient vector
+      ASSERT(l_ <= l_max_ && m_ <= l_max_, "coefficients iterator overflow");
+      size_t offset = ((3 + 2 * l_max_ - m_) * m_) / 2 + l_ - m_;
+      return LibsharpCoefficientInfo{
+          offset,
+          offset +
+              Spectral::Swsh::size_of_libsharp_coefficient_vector(l_max_) / 2,
+          l_max_, l_, m_};
+    }
+    /// advance the iterator by one position (prefix)
+    CoefficientsConstIterator& operator++() noexcept {
```

Comment:
Can you add the `operator++(int)` and `operator--(int)` variants as well for completeness? I can see someone (like myself) expecting to be able to do both `++it` and `it++`

Reply:
@markscheel and @fmahebert suggested removing the postfix operators I originally had in this PR under the philosophy that because they cause copies, it's better to present only the interface that one would suggest for current use-cases. If you feel strongly about following the standard, though, I can put them back.

Reply:
If copying 3 ints is what kills us I'll celebrate :D I'd rather have a conforming interface. There are cases where the postfix is actually what is preferred, i.e. one wants the copy. I know @fmahebert doesn't like people doing that, but I think it makes more sense to have a predictable interface.

Reply:
Fine 😛